[gn] introduce build_root variable

LuDuda opened the pull request:
#### Problem
The `build` folder has to be always present in the GN build root directory which in some cases requires creating a symbolic link for it (which is not available under Windows).

 #### Summary of Changes
This PR:
- Introduce `build_root` variable and build.gni file.
- Modifies the criteria of importing `ot_br_posix.gni` file.
- Modifies the output dir for shell library,
- Defines a custom platform's `build_overrides` files for nrfconnect.

rwalker-apple reviewed (approved):
LGTM, want @mspang to weigh in

LuDuda commented on config/nrfconnect/chip-gn/build_overrides/chip.gni:
```diff
@@ -0,0 +1,18 @@
+# Copyright (c) 2020 Project CHIP Authors
```
Fixed, and rebased on master.

LuDuda commented:
Rebased, and introduced a `build_root` variable and `build.gni` file after getting feedback from @mspang.
Additionally, bumped pigweed version, so now it is possible to build firmware under Windows OS as well.

Damian-Nordic reviewed (approved):
I approve, but please fix that one issue.

LuDuda commented on config/qpg6100/toolchain/BUILD.gn:
```diff
@@ -12,9 +12,10 @@
 # limitations under the License.
 
-import("//build_overrides/chip.gni")
+import("//build_overrides/build.gni")
```
Thanks, fixed.

mspang commented on config/nrfconnect/chip-gn/build_overrides/chip.gni:
```diff
@@ -0,0 +1,18 @@
+
+declare_args() {
+  # Root directory for CHIP.
+  chip_root = rebase_path("//../../..")
```
We should avoid paths like ("//.."). Such a path means that the root is incorrect because it does not contain all of the code. Absolute paths make caching less effective.

I might rather entirely remove stuff building in subdirectories than use absolute paths. Building in subdirectories was meant to provide an example.. this doesn't set a great one.

It might be OK if we restrict it to Windows and keep the symlinks and relative paths for other systems. Or we could do away with the separate .gn files and just script the CI system to create a new repository with CHIP as a submodule and test that it builds. Such an approach would work on Windows without contortions like /..

LuDuda replied:
Changed approach to build from the chip root. So we can resolve this comment.

mspang commented on build/config/BUILDCONFIG.gn:
```diff
@@ -12,6 +12,8 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import("//build_overrides/build.gni")
```
This introduces the contents of build.gni into the global scope. We need to avoid this because it makes it harder to integrate CHIP into other GN builds. That's why we imported these inside the private scope `(_chip_overrides`). I'd just move this import back there.

BUILDCONFIG.gn is the only file where this is a concern.

LuDuda replied:
Fixed now.

LuDuda commented:
New changes:
- Rebas to master
- Bump pigweed and GN versions to support `--root-target` option and building under Windows
- Build nrfconnect platform from CHIP root directory

Note that bumping pigweed will not casue problems like with #4490.

Damian-Nordic commented on config/nrfconnect/chip-gn/.gn:
```diff
@@ -22,5 +25,5 @@ default_args = {
   target_os = "zephyr"
 
-  import("//args.gni")
+  import("${chip_root}/config/nrfconnect/chip-gn/args.gni")
```
nit: if I read the GN documentation correctly `import("args.gni")` should be enough.

LuDuda replied:
This does not work with `--dotfile` argument.

Damian-Nordic commented on config/nrfconnect/chip-module/CMakeLists.txt:
```diff
@@ -170,10 +176,22 @@ if (CONFIG_CHIP_OPENTHREAD_CONFIG)
 # ==============================================================================
 
+if (CONFIG_CHIP_PW_RPC)
+    chip_gn_arg_import("${GN_ROOT_TARGET}/lib/pw_rpc/pw_rpc.gni")
```
fiy: This should be removed anyway once pigweed-app is refactored to import pw_rpc in CMake directly.

andy31415 commented:
@woody-apple @bzbarsky-apple - could you add the checkmark from Apple side?
